core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java

```diff
@@ -128,13 +138,109 @@ public Sep31PostTransactionResponse postTransaction(
 
     return Sep31PostTransactionResponse.builder()
         .id(txn.getId())
-        .stellarAccountId(asset.getDistributionAccount())
+        .stellarAccountId(Context.get().getAsset().getDistributionAccount())
         .stellarMemo(txn.getStellarMemo())
         .stellarMemoType(txn.getStellarMemoType())
         .build();
   }
 
-  private void generateTransactionMemo(Sep31Transaction txn) throws SepException {
+  void updateAmounts() throws AnchorException {
+    Sep31Transaction txn = Context.get().getTransaction();
+    Sep31PostTransactionRequest request = Context.get().getRequest();
+    Amount feeResponse = Context.get().getFee();
+
+    boolean strictSend = sep31Config.getPaymentType() == STRICT_SEND;
+    boolean withoutQuote = request.getQuoteId() == null;
+
+    AssetInfo reqAsset = assetService.getAsset(request.getAssetCode(), request.getAssetIssuer());
+    BigDecimal reqAmount = decimal(request.getAmount(), reqAsset);
+    BigDecimal fee = decimal(feeResponse.getAmount(), reqAsset);
+    String feeAsset = String.valueOf(feeResponse.getAsset());
+
+    BigDecimal amountIn;
+    String amountInAsset = reqAsset.getCode();
+    BigDecimal amountOut;
+    String amountOutAsset;
+
+    if (withoutQuote) {
+      // Without quote.
+      amountOutAsset = amountInAsset;
+      if (strictSend) {
+        // amount_in = req.amount
+        // amount_out = amount_in - amount fee
+        amountIn = reqAmount;
+        amountOut = amountIn.subtract(fee);
+      } else {
+        // amount_in = req.amount + fee
+        // amount_out = req.amount
+        amountIn = reqAmount.add(fee);
+        amountOut = reqAmount;
+      }
+    } else {
+      // With quote
+      Sep38Quote quote = sep38QuoteStore.findByQuoteId(request.getQuoteId());
+      txn.setQuoteId(request.getQuoteId());
+      amountOutAsset = quote.getBuyAsset();
+      BigDecimal buyAmount = decimal(quote.getBuyAmount(), reqAsset);
+      BigDecimal sellAmount = decimal(quote.getSellAmount(), reqAsset);
+
+      if (feeResponse.getAsset().equals(quote.getSellAsset())) {
+        // Fee is applied before the quote
+        if (strictSend) {
+          // amount_in=req.amount
+          // amount_out=quote.buy_amount
+          amountIn = reqAmount;
+          amountOut = buyAmount;
+          // Validate:
+          //   quote.sell_amount = amount_in - fee
+          if (sellAmount.compareTo(amountIn.subtract(fee)) != 0) {
+            throw new BadRequestException(
+                "Bad quote. quote.sell_amount does not equal to (amount_in - amount_fee)");
+          }
```

Comment:
This is tricky, right? This is part of why we're having these discussions 😥

Reply:
Yeah.... hope we can take the "Quote includes fee and tax" approach. Then we don't have these tricky calculations.